test/hotspot/gtest/metaspace/test_clms.cpp

```diff
@@ -0,0 +1,413 @@
+
+      MetaBlock bl(p, word_size);
+
+      if (is_class) {
+
+        EXPECT_TRUE(bl.is_aligned_base(_klass_arena_alignment_words));
+
+        if (_num_allocations == 1) {
+          // first allocation: nonclass arena unchanged, class arena grows by 1 chunk and wordsize,
+          // class arena freeblocks unchanged
+          EXPECT_ARENA_UNCHANGED(nca);
+          EXPECT_FREEBLOCKS_UNCHANGED(nca);
+          EXPECT_EQ(d_ca.num_chunks_delta, 1);
+          EXPECT_EQ((size_t)d_ca.used_words_delta, word_size);
+          EXPECT_FREEBLOCKS_UNCHANGED(ca);
+          return bl;
+        }
+
+        // Had this been taken from class arena freeblocks?
+        if (d_ca.num_freeblocks_delta == -1) {
+          // the class arena freeblocks should have gone down, and the non-class arena freeblocks may have gone
+          // up in case the block was larger than required
+          const size_t wordsize_block_taken = (size_t)(-d_ca.freeblocks_words_delta);
+          EXPECT_GE(wordsize_block_taken, word_size); // the block we took must be at least allocation size
+          const size_t expected_freeblock_remainder = wordsize_block_taken - word_size;
+          if (expected_freeblock_remainder > 0) {
+            // the remainder, if it existed, should have been added to nonclass freeblocks
+            EXPECT_EQ(d_nca.num_freeblocks_delta, 1);
+            EXPECT_EQ((size_t)d_nca.freeblocks_words_delta, expected_freeblock_remainder);
+          }
+          // finally, nothing should have happened in the arenas proper.
+          EXPECT_ARENA_UNCHANGED(ca);
+          EXPECT_ARENA_UNCHANGED(nca);
+          return bl;
+        }
+
+        // block was taken from class arena proper
+
+        // We expect allocation waste due to alignment, should have been added to the freeblocks
+        // of nonclass arena. Allocation waste can be 0. If no chunk turnover happened, it must be
+        // smaller than klass alignment, otherwise it can get as large as a commit granule.
+        const size_t max_expected_allocation_waste =
+            d_ca.num_chunks_delta == 0 ? (_klass_arena_alignment_words - 1) : Settings::commit_granule_words();
+        EXPECT_GE(d_ca.num_chunks_delta, 0);
+        EXPECT_LE(d_ca.num_chunks_delta, 1);
+        EXPECT_GE((size_t)d_ca.used_words_delta, word_size);
+        EXPECT_LE((size_t)d_ca.used_words_delta, word_size + max_expected_allocation_waste);
+        EXPECT_FREEBLOCKS_UNCHANGED(ca);
+        EXPECT_ARENA_UNCHANGED(nca);
+        if (max_expected_allocation_waste > 0) {
+          EXPECT_GE(d_nca.num_freeblocks_delta, 0);
+          EXPECT_LE(d_nca.num_freeblocks_delta, 1);
+          EXPECT_GE(d_nca.freeblocks_words_delta, 0);
+          EXPECT_LE((size_t)d_nca.freeblocks_words_delta, max_expected_allocation_waste);
+        } else {
+          EXPECT_FREEBLOCKS_UNCHANGED(nca);
+        }
+        return bl;
+
+      } // end: is_class
+
+      else
+
+      {
+        // Nonclass arena allocation.
```

Comment:
The style in this source file isn't really up to scratch, especially *these* lines. Anyway, it's in the tests, so I'm OK with this being fixed in a follow up RFE.

Reply:
Okay, will fix